File: Frontend/src/types/prediction.ts
export type AttendanceRecord = {
  date: string;
  status: 'PRESENT' | 'ABSENT' | 'LATE' | 'EXCUSED';
  session_name: string;
  notes?: string;
};

export type StudentInput = {
  attendance: number;
  assignment_completion: number;
  test_average: number;
  coding_hours: number;
  goals_completed: number;
  projects_completed: number;
  interview_practice_hours: number;
};

export type StudentProfile = StudentInput & {
  id: string;
  name: string;
  email: string;
  prediction: PredictionLabel;
  attendance_records: AttendanceRecord[];
};

export type PredictionLabel = 'AT_RISK' | 'NEEDS_ATTENTION' | 'ON_TRACK';

export type PredictionProbabilities = {
  at_risk: number;
  needs_attention: number;
  on_track: number;
};

export type RAGDocument = {
  id: string;
  category: string;
  title: string;
  description: string;
  relevance_score: number;
  action_items: string[];
};

export type RAGActionPlanDay = {
  day: number;
  task: string;
};

export type RAGResource = {
  name: string;
  url: string;
  type: string;
  source_doc?: string;
};

export type RAGRecommendation = {
  risk_drivers: string[];
  retrieved_documents: RAGDocument[];
  action_plan: RAGActionPlanDay[];
  resources: RAGResource[];
  rag_metadata: {
    engine: string;
    indexed_documents_count: number;
    top_k_retrieved: number;
  };
};

export type PredictionResponse = {
  prediction: PredictionLabel;
  probabilities: PredictionProbabilities;
  is_temporary?: boolean;
  note?: string;
  rag_recommendations?: RAGRecommendation;
};

File: Backend/app/schemas/prediction.py
"""Pydantic schemas for prediction responses and RAG recommendations."""

from __future__ import annotations
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class PredictionProbabilities(BaseModel):
    at_risk: float = Field(..., ge=0, le=1)
    needs_attention: float = Field(..., ge=0, le=1)
    on_track: float = Field(..., ge=0, le=1)


class RAGDocumentSchema(BaseModel):
    id: str
    category: str
    title: str
    description: str
    relevance_score: float
    action_items: List[str] = []


class RAGActionPlanDaySchema(BaseModel):
    day: int
    task: str


class RAGResourceSchema(BaseModel):
    name: str
    url: str
    type: str
    source_doc: Optional[str] = None


class RAGRecommendationSchema(BaseModel):
    risk_drivers: List[str]
    retrieved_documents: List[RAGDocumentSchema]
    action_plan: List[RAGActionPlanDaySchema]
    resources: List[RAGResourceSchema]
    rag_metadata: Dict[str, Any]


class PredictionResponse(BaseModel):
    prediction: str
    probabilities: PredictionProbabilities
    is_temporary: bool = False
    note: str = "Machine Learning Inference & RAG Vector Retrieval Complete."
    rag_recommendations: Optional[RAGRecommendationSchema] = None